test/hotspot/gtest/runtime/test_os.cpp

```diff
@@ -405,6 +405,61 @@ TEST_VM(os, jio_snprintf) {
   test_snprintf(jio_snprintf, false);
 }
 
+#ifndef MAX_PATH
+#define MAX_PATH    (2 * K)
+#endif
+
+TEST_VM(os, realpath) {
+  /* POSIX requires that the file exists, Windows doesn't */
+  static const char* nosuchpath = "/1234567890123456789";
+  static const char* tmppath = "/tmp";
+
+  char buffer[MAX_PATH];
+
+  errno = 0;
+  const char* returnedBuffer = os::realpath(nosuchpath, buffer, sizeof(nosuchpath) - 2);
+  /* Returns ENOENT on Linux, ENAMETOOLONG on Windows */
+  EXPECT_TRUE(returnedBuffer == nullptr);
+  EXPECT_TRUE(errno == ENAMETOOLONG || errno == ENOENT);
+
+  errno = 0;
+  returnedBuffer = os::realpath(nosuchpath, buffer, sizeof(nosuchpath) + 3);
+  /* Returns ENOENT on Linux, 0 on Windows */
+  EXPECT_TRUE(errno == 0 || errno == ENOENT);
+  EXPECT_TRUE(returnedBuffer == nullptr || returnedBuffer == buffer);
+
+  errno = 0;
+  returnedBuffer = os::realpath(tmppath, buffer, MAX_PATH);
+  EXPECT_TRUE(returnedBuffer == buffer);
+
+  errno = 0;
+  returnedBuffer = os::realpath(tmppath, buffer, strlen(tmppath) + 3);
+  EXPECT_TRUE(returnedBuffer == buffer);
+
+  errno = 0;
+  returnedBuffer = os::realpath(tmppath, buffer, strlen(tmppath) - 1);
+  EXPECT_TRUE(errno == ENAMETOOLONG);
+  EXPECT_TRUE(returnedBuffer == nullptr);
+
+  /* the following tests cause an assert in fastdebug mode */
+  DEBUG_ONLY(if (false)) {
+    errno = 0;
+    returnedBuffer = os::realpath(nullptr, buffer, sizeof(buffer));
+    EXPECT_TRUE(errno == EINVAL);
+    EXPECT_TRUE(returnedBuffer == nullptr);
+
+    errno = 0;
+    returnedBuffer = os::realpath(tmppath, buffer, sizeof(buffer));
+    EXPECT_TRUE(errno == EINVAL);
```

Comment:
How is this an EINVAL case?

Reply:
You are correct - this was a cut and paste error; the call tested should have been
`os::realpath(tmppath, nullptr, sizeof(buffer));`